Source/WebCore/platform/graphics/ImageBuffer.cpp

```diff
@@ -171,4 +201,41 @@ void ImageBuffer::drawConsuming(RefPtr<ImageBuffer> imageBuffer, GraphicsContext
     imageBuffer->drawConsuming(context, destRect, srcRect, options);
 }
 
+String ImageBuffer::toDataURL(const String& mimeType, std::optional<double> quality, PreserveResolution preserveResolution) const
+{
+    return toDataURL(Ref { const_cast<ImageBuffer&>(*this) }, mimeType, quality, preserveResolution);
+}
+
+Vector<uint8_t> ImageBuffer::toData(const String& mimeType, std::optional<double> quality, PreserveResolution preserveResolution) const
+{
+    return toData(Ref { const_cast<ImageBuffer&>(*this) }, mimeType, quality, preserveResolution);
+}
+
+String ImageBuffer::toDataURL(Ref<ImageBuffer> source, const String& mimeType, std::optional<double> quality, PreserveResolution preserveResolution)
+{
+    auto encodedData = toData(WTFMove(source), mimeType, quality, preserveResolution);
+    if (encodedData.isEmpty())
+        return "data:,"_s;
+    return makeString("data:", mimeType, ";base64,", base64Encoded(encodedData));
+}
+
+Vector<uint8_t> ImageBuffer::toData(Ref<ImageBuffer> source, const String& mimeType, std::optional<double> quality, PreserveResolution preserveResolution)
+{
+    RefPtr<NativeImage> image;
+    if (!platformMIMETypeIsJPEG(mimeType))
+        image = copyNativeImageImpl(WTFMove(source), DontCopyBackingStore, preserveResolution);
+    else {
+        // Composite this ImageBuffer on top of opaque black, because JPEG does not have an alpha channel.
+        auto copyBuffer = copyImageBuffer(WTFMove(source), preserveResolution);
+        if (!copyBuffer)
+            return { };
+        // We composite the copy on top of black by drawing black under the copy.
+        copyBuffer->context().fillRect({ { }, copyBuffer->logicalSize() }, Color::black, CompositeOperator::DestinationOver);
+        image = sinkIntoNativeImage(WTFMove(copyBuffer));
+    }
+    if (!image)
+        return { };
+    return data(image->platformImage().get(), mimeType, quality);
```

Comment:
Not new. But data() is a little bit generic name and does not reveal what it is doing. Can we use a better name like encodeToData() or something similar?

Reply:
Changed to encodeData().